src/soc/src/sunxi/d1/smhc/card.rs

```diff
@@ -0,0 +1,98 @@
+/// The different types of cards
+#[derive(Debug, PartialEq)]
+pub enum CardType {
+    SD1,
+    SD2,
+    SDHC,
+}
+
+/// Card status in R1 response format
+pub struct CardStatus(u32);
+
+#[derive(Debug, PartialEq)]
+pub enum CardCurrentState {
+    Idle,
+    Ready,
+    Ident,
+    Stby,
+    Tran,
+    Data,
+    Rcv,
+    Prg,
+    Dis,
+    Reserved,
+}
+
+impl From<u32> for CardCurrentState {
+    fn from(value: u32) -> Self {
+        match value {
+            0 => Self::Idle,
+            1 => Self::Ready,
+            2 => Self::Ident,
+            3 => Self::Stby,
+            4 => Self::Tran,
+            5 => Self::Data,
+            6 => Self::Rcv,
+            7 => Self::Prg,
+            8 => Self::Dis,
+            _ => Self::Reserved,
+        }
+    }
+}
+
+impl CardStatus {
+    pub(crate) fn new(val: u32) -> Self {
+        Self(val)
+    }
+
+    /// Indicates if there was an error in the sequence of the authentication process.
+    pub fn auth_seq_error(&self) -> bool {
+        const BIT: u32 = 1 << 3;
+        (self.0 & BIT) == BIT
```

Comment:
A lot of times, in firmware, when things to wrong, you want to have reasonable info. Is there additional information in self.0 that would be useful to have available to the caller when things have gone wrong?

Reply:
In the public interface for `CardStatus` I added functions for accessing the bits
that I thought could be interesting. It's not complete, but functions can be added as needed.
Or I can add a `pub fn raw(&self) -> u32` or just make `self.0` pub.

<details>
  <summary>List of card status bits</summary>
  
  ```
  Bit  Typ Clr Identifier         Meaning
  31   ERX  C  OUT_OF_RANGE       (1=Command's argument was out of range)
  30   ERX  C  ADDRESS_ERROR      (1=Misaligned address/block len mismatch)
  29   ERX  C  BLOCK_LEN_ERROR    (1=Wrong block length, bytelen mismatch)
  28   ER   C  ERASE_SEQ_ERROR    (1=Error in erase command sequence)
  27   ERX  C  ERASE_PARAM        (1=Wrong erease selection of write-blocks)
  26   ERX  C  WP_VIOLATION       (1=Write failed due to write-protection)
  25   SX   A  CARD_IS_LOCKED     (1=Card is locked by the host)
  24   ERX  C  LOCK_UNLOCK_FAILED (1=Lock/unlock sequence or password error)
  23   ER   B  COM_CRC_ERROR      (1=CRC check of previous command failed)
  22   ER   B  ILLEGAL_COMMAND    (1=Command not legal for the card state)
  21   ERX  C  CARD_ECC_FAILED    (1=Internal error correction failed)
  20   ERX  C  CC_ERROR           (1=Internal card controller error)
  19   ERX  C  ERROR              (1=General error, or Unknown error)
  18   -    -  Reserved (eMMC: UNDERRUN)
  17   -    -  Reserved (eMMC: OVERRUN) (eSD: DEFERRED_RESPONSE)
  16   ERX  C  CSD_OVERWRITE      (1=read-only CSD section doesn't match card
                                  content, or attempted to reverse the
                                  Copy/WP bits)
  15   ERX  C  WP_ERASE_SKIP      (1=partial erase error due to write-protect)
  14   SX   A  CARD_ECC_DISABLED  (1=Internal error correction wasn't used)
  13   SR   C  ERASE_RESET        (1=Erase sequence was aborted)
  12-9 SX   B  CURRENT_STATE      (00h..0Fh=state, see below)
  8    SX   A  READY_FOR_DATA     (1=Ready/buffer is empty)
  7    EX   C  SWITCH_ERROR       (1=SWITCH command refused, MMC only)
  6    -    -  Reserved/Unspecified (description is left blank)
  5    SR   C  APP_CMD            (1=Card will expect ACMD)
  4    -    -  Reserved for SD I/O Card
  3    ER   C  AKE_SEQ_ERROR      (1=Authentication Sequence Error)
  2    -    -  Reserved for application specific commands
  1-0  -    -  Reserved for manufacturer test mode
  ```
</details>